query-engine/core/src/schema_builder/input_types/fields/data_input_mapper/update.rs

```diff
@@ -203,3 +196,119 @@ fn update_operations_object_type(
 
     Arc::downgrade(&obj)
 }
+
+// Builds a `set` input object type. Should be used for the shorthand syntax and within the update envelope type.
+fn composite_set_update_object_type(ctx: &mut BuilderContext, cf: &CompositeFieldRef) -> InputObjectTypeWeakRef {
+    let name = format!("{}SetUpdateInput", cf.typ.name);
+
+    let ident = Identifier::new(name, PRISMA_NAMESPACE);
+    return_cached_input!(ctx, &ident);
+
+    let input_object = Arc::new(init_input_object_type(ident.clone()));
+    ctx.cache_input_type(ident, input_object.clone());
+
+    let mapper = CreateDataInputFieldMapper::new_checked();
+    let fields = mapper.map_all(ctx, cf.typ.fields());
+
+    input_object.set_fields(fields);
+
+    Arc::downgrade(&input_object)
+}
+
+/// Build an operation envelope object type for composite updates.
+/// An operation envelope is an object that encapsulates the possible operations, like:
+/// ```text
+/// cf_field: { // this is the envelope object
+///   set:    { ... set type ... }
+///   update: { ... update type ... }
+///   ... more ops ...
+/// }
+/// ```
+fn composite_update_envelope_object_type(ctx: &mut BuilderContext, cf: &CompositeFieldRef) -> InputObjectTypeWeakRef {
+    let name = format!("{}UpdateEnvelopeInput", cf.typ.name);
+
+    let ident = Identifier::new(name, PRISMA_NAMESPACE);
+    return_cached_input!(ctx, &ident);
+
+    let mut input_object = init_input_object_type(ident.clone());
+    input_object.require_exactly_one_field();
+
+    let input_object = Arc::new(input_object);
+    ctx.cache_input_type(ident, input_object.clone());
+
+    let mut fields = vec![composite_set_update_input_field(ctx, cf)];
+
+    append_opt(&mut fields, composite_update_input_field(ctx, cf));
+    append_opt(&mut fields, composite_push_update_input_field(ctx, cf));
+    append_opt(&mut fields, composite_unset_update_input_field(cf));
+
+    input_object.set_fields(fields);
+
+    Arc::downgrade(&input_object)
+}
+
+/// Builds the `update` input object type. Should be used in the envelope type.
+fn composite_update_object_type(ctx: &mut BuilderContext, cf: &CompositeFieldRef) -> InputObjectTypeWeakRef {
+    let name = format!("{}UpdateInput", cf.typ.name);
+
+    let ident = Identifier::new(name, PRISMA_NAMESPACE);
+    return_cached_input!(ctx, &ident);
+
+    let mut input_object = init_input_object_type(ident.clone());
+    input_object.set_min_fields(1);
```

Comment:
Do we agree on the `set_min_fields(1)` for this object type? This is to prevent:

```graphql
mutation { updateOneX(data: {
  cf_field: { update: {} <-- this empty object }
}) } {
  id
}
```

Reply:
A quick experiment shoes me that we don't do that for other updates as well. Is this important? We could just treat it as noop like relations.

Reply:
Yeah, I noticed it wasn't the case on existing `update`s. I thought this might remove potential ambiguities where users might think `{ a: { update: {} } }` would update `a` to `{}`

Reply:
Decision: keep the `min(1)`.